Double randomizer mode

hatkirby opened the pull request:
This mode removes some impossible situations that may arise when using this randomizer after using Sigma's randomizer:

- Tutorial Back Left is shuffled within a smaller pool
- Treehouse pivots only shuffle amongst themselves
- Swamp Entry is not shuffled
- Mountain metapuzzle interior panels are not shuffled
- Symmetry Laser Yellows and Blues only shuffle amongst themselves
- UTM Perspective puzzles only shuffle amongst themselves

hatkirby committed:
Added checkbox for double randomizer mode

hatkirby committed:
Prevented Tutorial Back Left from being impossible in double randomizer mode

hatkirby committed:
Prevented mountain metapuzzle from being impossible in double randomizer mode

hatkirby committed:
Prevented Swamp Entry from being impossible in double randomizer mode

hatkirby committed:
Moved the comment about Tutorial Back Left so it's with the other comments

hatkirby committed:
Prevented Treehouse pivots from being impossible in double randomizer mode

hatkirby committed:
Resized label for double randomizer mode

hatkirby committed:
Reworked double randomizer mode so that it can be disabled after being used

hatkirby committed:
Randomized the puzzles that double randomizer mode controls for before anything else

hatkirby committed:
Prevented shuffling Symmetry Laser Yellow and Blue panels in double randomizer mode

hatkirby committed:
Symmetry Laser Yellows and Blues are shuffled amongst themselves in double randomizer mode

hatkirby committed:
UTM perspective puzzles are only shuffled amongst themselves in double randomizer mode

hatkirby committed:
Combined the various _doubleRandomizer if's in Randomize()

hatkirby committed:
Renamed a couple of panel sets for clarity

jbzdarkid commented on Source/Randomizer.cpp:
```diff
@@ -132,16 +132,50 @@ void Randomizer::Randomize() {
+        Randomize(upDownPanelsSetOneDoubleMode, SWAP::LINES | SWAP::COLORS);
+        Randomize(upDownPanelsSetTwoDoubleMode, SWAP::LINES | SWAP::COLORS);
+        Randomize(upDownPanelsSetFour, SWAP::LINES | SWAP::COLORS);
+        // Many puzzles either crash the game or do not solve properly when
+        // swapped with Swamp Entry. To make things simpler, we will just remove
+        // that panel from both pools it is found in.
+        Randomize(quarryLaserOptionsDoubleMode, SWAP::LINES | SWAP::COLORS);
+        Randomize(squarePanelsDoubleMode, SWAP::LINES | SWAP::COLORS);
```
For simple modifications like this, can we just remove the dangerous panels? That way we don't have to maintain two similar (but long) panel lists.
```suggestion
        // Many puzzles either crash the game or do not solve properly when
        // swapped with Swamp Entry. To make things simpler, we will just remove
        // that panel from both pools it is found in.
        quarryLaserOptions.remove(0x00A52); squarePanels.remove(0x00A52), // Symmetry Island Laser Yellow 1
        quarryLaserOptions.remove(0x00A57); squarePanels.remove(0x00A57), // Symmetry Island Laser Yellow 2
        quarryLaserOptions.remove(0x00A5B); squarePanels.remove(0x00A5B), // Symmetry Island Laser Yellow 3
        quarryLaserOptions.remove(0x00A61); squarePanels.remove(0x00A61), // Symmetry Island Laser Blue 1
        quarryLaserOptions.remove(0x00A64); squarePanels.remove(0x00A64), // Symmetry Island Laser Blue 2
        quarryLaserOptions.remove(0x00A68); squarePanels.remove(0x00A68), // Symmetry Island Laser Blue 3
        quarryLaserOptions.remove(0x0056E); squarePanels.remove(0x0056E), // Swamp Entry
        Randomize(quarryLaserOptionsDoubleMode, SWAP::LINES | SWAP::COLORS);
        Randomize(squarePanelsDoubleMode, SWAP::LINES | SWAP::COLORS);
```

hatkirby replied:
The reason I didn't do this was because it'll mess things up if a player uses double randomizer mode, then starts a new game and uses the randomizer again with double randomizer mode unchecked. We could still do something like this though by copying each set into a local variable, removing the dangerous panels, and then using those sets for randomizing. I agree that it would be better than duplicating the huge lists.

hatkirby replied:
Changed & tested (for this and for the other comments).

jbzdarkid commented on Source/Randomizer.cpp:
```diff
@@ -132,16 +132,50 @@ void Randomizer::Randomize() {
+        // that panel from both pools it is found in.
+        Randomize(quarryLaserOptionsDoubleMode, SWAP::LINES | SWAP::COLORS);
+        Randomize(squarePanelsDoubleMode, SWAP::LINES | SWAP::COLORS);
+    }
+    else {
```
```suggestion
    } else {
```
[nit] style in this codebase

jbzdarkid commented on Source/Randomizer.cpp:
```diff
@@ -132,16 +132,50 @@ void Randomizer::Randomize() {
+        // shuffle these six panels amongst themselves.
+        Randomize(symmetryLaserYellows, SWAP::LINES | SWAP::COLORS);
+        Randomize(symmetryLaserBlues, SWAP::LINES | SWAP::COLORS);
+        // Prevent the fourth UTM perspective puzzle from being shuffled, as it
+        // is unknown whether any of the other puzzles in these pools can be
+        // solved there after using Sigma's randomizer.
+        Randomize(upDownPanelsSetZeroDoubleMode, SWAP::LINES | SWAP::COLORS);
+        Randomize(upDownPanelsSetOneDoubleMode, SWAP::LINES | SWAP::COLORS);
+        Randomize(upDownPanelsSetTwoDoubleMode, SWAP::LINES | SWAP::COLORS);
+        Randomize(upDownPanelsSetFour, SWAP::LINES | SWAP::COLORS);
```
Same suggestion as below -- if it's a single (or small subset) of bad panels, let's just remove them.
```suggestion
        // Prevent the fourth UTM perspective puzzle from being shuffled, as it
        // is unknown whether any of the other puzzles in these pools can be
        // solved there after using Sigma's randomizer.
        upDownPanelsSetZero.remove(0x288AA); // UTM Perspective 4
        upDownPanelsSetOne.remove(0x288AA); // UTM Perspective 4
        upDownPanelsSetTwo.remove(0x288AA); // UTM Perspective 4
        Randomize(upDownPanelsSetZero, SWAP::LINES | SWAP::COLORS);
        Randomize(upDownPanelsSetOne, SWAP::LINES | SWAP::COLORS);
        Randomize(upDownPanelsSetTwo, SWAP::LINES | SWAP::COLORS);
        Randomize(upDownPanelsSetFour, SWAP::LINES | SWAP::COLORS);
```

jbzdarkid commented on Source/Randomizer.cpp:
```diff
@@ -132,16 +132,50 @@ void Randomizer::Randomize() {
+        // be solveable in the down position after using Sigma's randomizer. We
+        // do not want to swap any such panel with Tutorial Back Left, because
+        // that would make it impossible to exit Tutorial.
+        Randomize(upDownPanelsSetThreeDoubleMode, SWAP::LINES | SWAP::COLORS);
```
I could go either way here, since these two lists are relatively different -- IMO it's more long-term maintainable to have a single list and modify it. Up to you.

hatkirby committed:
Removed DoubleMode duplicates of squarePanels and quarryLaserOptions

hatkirby committed:
Removed DoubleMode duplicates of upDownPanels sets 0-2

hatkirby committed:
Removed DoubleMode duplicate of upDownPanelsSetThree

hatkirby committed:
Style stuff

hatkirby committed:
Compile errors

jbzdarkid commented on Source/Randomizer.cpp:
```diff
@@ -109,6 +111,14 @@ int find(const std::vector<T> &data, T search, size_t startIndex = 0) {
     throw std::exception("Couldn't find value in data!");
 }
 
+std::vector<int> copyWithoutElements(const std::vector<int>& input, const std::set<int>& toRemove) {
+  std::vector<int> result = input;
+  result.erase(std::remove_if(std::begin(result), std::end(result), [&toRemove](int val) {
+      return toRemove.count(val);
+  }), std::end(result));
+  return result;
+}
```
Good idea, but the implementation here is a bit odd.
- Why not use an in/out variable? We don't need the input list to be immutable, and copying is potentially slow.
- There's a function called `std::remove_copy_if` which I feel like is more suited to the job, if you keep this orchestration.
- As wonderful as std::algorithm is, this code is now very hard to read (and reason about). Why not just use a simple for loop?
```cpp
for (auto& it : input) {
    if (toRemove.find(it) != toRemove.end()) it = input.remove(it);
}
```

hatkirby replied:
I think we do want to do a copy, because otherwise we're modifying the original set of panels, and that modification will be carried forward if the player later unchecks "double randomizer mode".
 
The code sample doesn't work because vector doesn't have a remove-by-value method, and even if it did, we'd be invalidating the iterator that the for loop is using. If we want to avoid using algorithm, I think the simplest thing might be something like this:

```c++
std::vector<int> result;
for (int val : input) {
    if (toRemove.find(val) == toRemove.end()) {
        result.push_back(val);
    }
}
```

jbzdarkid replied:
Whoops, I confused `remove` with `erase` again. Anyways, that's why it's `it = input.erase(it);`, erase returns a new iterator which will work safely with the for loop.

That said, if we're making a copy, that's not relevat.
Your suggestion looks good, just call `result.reserve()` to avoid extra resizing.

hatkirby replied:
Sounds good to me!

jbzdarkid commented on Source/Randomizer.cpp:
```diff
@@ -132,16 +142,75 @@ void Randomizer::Randomize() {
+        // be solveable in the down position after using Sigma's randomizer. We
+        // do not want to swap any such panel with Tutorial Back Left, because
+        // that would make it impossible to exit Tutorial.
+        const std::set<int> bannedTutorialBackLeftPanels = {
+            0x00070, // Symmetry Island Fading Lines 5
+            0x01E5A, // Mill Entry Door Left
+            0x00072, // Symmetry Island Fading 3
+            0x00076, // Symmetry Island Fading 7
+            0x3C125, // Mill Control Room Extra Panel
+            0x09E85, // Tunnels Into Town Big Triangles
+        };
+        std::vector<int> upDownPanelsSetThreeDoubleMode = copyWithoutElements(upDownPanelsSetThree, bannedTutorialBackLeftPanels);
```
If you're not re-using the variable, just inline it.
```suggestion
        std::vector<int> upDownPanelsSetThreeDoubleMode = copyWithoutElements(upDownPanelsSetThree, {
            0x00070, // Symmetry Island Fading Lines 5
            0x01E5A, // Mill Entry Door Left
            0x00072, // Symmetry Island Fading 3
            0x00076, // Symmetry Island Fading 7
            0x3C125, // Mill Control Room Extra Panel
            0x09E85, // Tunnels Into Town Big Triangles
        });
```

jbzdarkid commented on Source/Randomizer.cpp:
```diff
@@ -132,16 +142,75 @@ void Randomizer::Randomize() {
+        // Prevent the fourth UTM perspective puzzle from being shuffled, as it
+        // is unknown whether any of the other puzzles in these pools can be
+        // solved there after using Sigma's randomizer.
+        const std::set<int> bannedUpDownPanels {
+            0x288AA, // UTM Perspective 4
+        };
+        std::vector<int> upDownPanelsSetZeroDoubleMode = copyWithoutElements(upDownPanelsSetZero, bannedUpDownPanels);
+        std::vector<int> upDownPanelsSetOneDoubleMode = copyWithoutElements(upDownPanelsSetOne, bannedUpDownPanels);
+        std::vector<int> upDownPanelsSetTwoDoubleMode = copyWithoutElements(upDownPanelsSetTwo, bannedUpDownPanels);
```
Even though you're re-using this variable, I don't think it really helps readability here to have a separate set. I would just add a comment and then focus on clean code.
```suggestion
        // 0x288AA UTM Perspective 4
        std::vector<int> upDownPanelsSetZeroDoubleMode = copyWithoutElements(upDownPanelsSetZero, {0x288AA});
        std::vector<int> upDownPanelsSetOneDoubleMode = copyWithoutElements(upDownPanelsSetOne, {0x288AA});
        std::vector<int> upDownPanelsSetTwoDoubleMode = copyWithoutElements(upDownPanelsSetTwo, {0x288AA});
```

jbzdarkid commented on Source/Randomizer.cpp:
```diff
@@ -132,16 +142,75 @@ void Randomizer::Randomize() {
+        // Many puzzles either crash the game or do not solve properly when
+        // swapped with Swamp Entry. To make things simpler, we will just remove
+        // that panel from both pools it is found in.
+        const std::set<int> bannedSquarePanels {
+            0x00A52, // Symmetry Laser Yellow 1
+            0x00A57, // Symmetry Laser Yellow 2
+            0x00A5B, // Symmetry Laser Yellow 3
+            0x00A61, // Symmetry Laser Blue 1
+            0x00A64, // Symmetry Laser Blue 2
+            0x00A68, // Symmetry Laser Blue 3
+            0x0056E, // Swamp Entry
+        };
```
This list I'm 100% on board with. In fact, I would encourage you to move it into `panels.h` -- alongside all our other long lists of panels.

hatkirby committed:
Inlined bannedTutorialBackLeftPanels

Co-authored-by: Joseph Blackman <jbzdarkid@users.noreply.github.com>

hatkirby committed:
Inlined bannedUpDownPanels

Co-authored-by: Joseph Blackman <jbzdarkid@users.noreply.github.com>

hatkirby committed:
Moved bannedSquarePanels to Panels.h

hatkirby committed:
Simpler implementation of copyWithoutElements

jbzdarkid commented on Source/Panels.h:
```diff
@@ -771,6 +772,16 @@ std::vector<int> squarePanels = {
     0x09E85, // Tunnels Town Shortcut
 };
 
+std::set<int> doubleModeBannedSquarePanels = {
```
```suggestion
std::vector<int> doubleModeBannedSquarePanels = {
```
There is no noticable perf benefit to set over vector when the data set is small. Let's use vector for consistency.

hatkirby replied:
So, I just tried this and apparently vector doesn't have a find method? We could either use algorithm's std::find, or switch back to set, whichever you prefer.

jbzdarkid replied:
std::find is ok by me.

jbzdarkid commented on Source/Panels.h:
```diff
@@ -1,5 +1,6 @@
 #pragma once
 #include <vector>
+#include <set>
```
```suggestion
```
(see other comment)

hatkirby committed:
Reserved space in output vector in copyWithoutElements

jbzdarkid commented on Source/Randomizer.cpp:
```diff
@@ -147,15 +147,14 @@ void Randomizer::Randomize() {
             0x00072, // Symmetry Island Fading 3
             0x00076, // Symmetry Island Fading 7
             0x3C125, // Mill Control Room Extra Panel
             0x09E85, // Tunnels Into Town Big Triangles
```
I think we called this panel Tunnels Town Shortcut? Not really picky, just wondering if the name is different elsewhere.

hatkirby replied:
In splits I usually see it as UTM Town Shortcut; this name is just the name I copied out of Panels.h.

Actually, looking into Panels.h, I see all three names being used in various places. I could change them all to be consistent if you want.

jbzdarkid replied:
I would love that, thank you. I don't know if we're using UTM or Tunnels in this project, but "Town Shortcut" is definitely the right name.

hatkirby committed:
Replaced set with vector in double randomizer code

hatkirby committed:
Standardized "Tunnels Town Shortcut" panel name

There is one "UTM Town Shortcut" that I didn't change because it had a different panel ID (and was commented out).

jbzdarkid commented:
Looks good -- only thing left is to bump version, I think.

hatkirby commented:
Should I do that or do you do it?

jbzdarkid commented:
> Should I do that or do you do it?

If there's anyone else contributing to this release I can do it once their changes are in -- if nobody else is writing code you should do it.

hatkirby committed:
Bump version

hatkirby commented:
Done.

jbzdarkid commented on App/Version.h:
```diff
@@ -4,7 +4,7 @@
 
 #define MAJOR 7
-#define MINOR 1
+#define MINOR 2
```
There's actually more to it than this. Leave it in, I'll do the follow-up work.
